py/packages/genkit/src/genkit/core/plugin_abc.py

```diff
@@ -8,72 +8,26 @@
 import abc
 import typing
 
-from genkit.core.schema_types import GenerateRequest, GenerateResponse
-
 if typing.TYPE_CHECKING:
+    # TODO: Must point to an abstraction, not an actual implementation,
+    #       which should resolve the circular import problem.
     from genkit.veneer import Genkit
 
 
 class Plugin(abc.ABC):
     """
-    Abstract class defining common interface
-    for the Genkit Plugin implementation
-
-    NOTE: Any plugin defined for the Genkit must inherit from this class
+    Abstract base class for plugins.
     """
 
     @abc.abstractmethod
-    def attach_to_veneer(self, veneer: Genkit) -> None:
+    def register(self, ai: Genkit) -> None:
         """
-        Entrypoint for attaching the plugin to the requested Genkit Veneer
-
-        Implementation must be provided for any inheriting plugin.
-
-        Args:
-            veneer: requested `genkit.veneer.Genkit` instance
-
-        Returns:
-            None
+        Entrypoint for attaching the plugin to the Genkit Veneer.
         """
         pass
 
-    def _add_model_to_veneer(
-        self, veneer: Genkit, name: str, metadata: dict | None = None
-    ) -> None:
-        """
-        Defines plugin's model in the Genkit Registry
-
-        Uses self._model_callback as a generic callback wrapper
-
-        Args:
-            veneer: requested `genkit.veneer.Genkit` instance
-            name: name of the model to attach
-            metadata: metadata information associated
-                      with the provided model (optional)
-
-        Returns:
-            None
-        """
-        if not metadata:
-            metadata = {}
-        veneer.define_model(
-            name=name, fn=self._model_callback, metadata=metadata
-        )
-
-    @abc.abstractmethod
-    def _model_callback(self, request: GenerateRequest) -> GenerateResponse:
```

Comment:
if we remove `_model_callback` then what was the point of refactor? we will end up with completely the same spaghetti code but with the class wrapper

Reply:
yes, we do, as I mentioned previously, purpose of the model callback is to define single entry point interface for the model action handlers. This will allow developer to clearly understand what are the items that he should implement when defining new Plugin class. In addition, if we use generic "attach to model" interface, we encapsulate calling of model_callback which will give us working solution with minimal code rewriting.